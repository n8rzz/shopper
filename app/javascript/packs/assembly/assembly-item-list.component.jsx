import React from 'react';
import PropTypes from 'prop-types';
import { AssemblyItemContainer } from './assembly-item.container';

export function AssemblyItemList(props) {
    return (
        <ul className={'vlist mix-vlist_isClickable js-accordionList'}>
            {props.assemblies.map((assembly, index) => (
                <AssemblyItemContainer
                    key={`assemblyItemContainer-${index}`}
                    assembly={assembly}
                    assemblyItemsGroupedByAssemblyId={props.assemblyItemsGroupedByAssemblyId}
                    csrf={props.csrf}
                    departmentIdMap={props.departmentIdMap}
                    itemIdMap={props.itemIdMap}
                />
            ))}
        </ul>
    );
}

AssemblyItemList.propTypes = {
    assemblies: PropTypes.array.isRequired,
    assemblyItemsGroupedByAssemblyId: PropTypes.object.isRequired,
    csrf: PropTypes.string.isRequired,
    departmentIdMap: PropTypes.object.isRequired,
    itemIdMap: PropTypes.object.isRequired,
};
