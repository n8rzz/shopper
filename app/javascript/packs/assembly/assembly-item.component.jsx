import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';

export class AssemblyItem extends React.Component {
    _buildAssemblyGroupHeaderJsx() {
        const assemblyItemList = this.props.assemblyItemsGroupedByAssemblyId[this.props.assembly.id];

        return (
            <div className={'listItem'}>
                <div className={'listItem-hd'}>
                    <h2
                        className={'hdg hdg_2'}
                        onClick={this.props.onClickAccordionTriggerHandler}
                    >
                        {this.props.assembly.name}
                    </h2>
                    <div>
                        <button
                            type={'button'}
                            data-assembly-id={this.props.assembly.id}
                            onClick={this.props.onClickAddAssemblyHandler}
                        >
                            Add to order
                        </button>
                    </div>
                </div>
                <div className={'listItem-ft'}>
                    <ul className={'stereo'}>
                        <li>
                            Items:
                            {assemblyItemList.length}
                        </li>
                        <li>
                            <a
                                className={'link'}
                                href={'/assemblies/'}
                            >
                                Edit Assembly
                            </a>
                        </li>
                    </ul>
                </div>
            </div>
        );
    }

    _buildAssemblyItemList() {
        const assemblyItemList = this.props.assemblyItemsGroupedByAssemblyId[this.props.assembly.id];
        const assemblyItemListJsx = assemblyItemList.map((assemblyItem, index) => {
            const item = this.props.itemIdMap[assemblyItem.item_id];
            const departmentName = this.props.departmentIdMap[item.department_id].name;

            return (
                <li
                    key={`assemblyItem-${index}`}
                    className={'listItem listItem-child'}
                >
                    <div className={'listItem-hd'}>
                        <h3 className={'hdg'}>
                            {item.name}
                        </h3>
                    </div>
                    <div className={'listItem-ft'}>
                        <ul className={'stereo'}>
                            <li>
                                {departmentName}
                            </li>
                            <li>
                                <form>
                                    <button
                                        type={'submit'}
                                        data-department-id={item.department_id}
                                        data-assembly-item-id={assemblyItem.id}
                                        onClick={this.props.onClickAddAssemblyItemHandler}
                                    >
                                        Add Item to order
                                    </button>
                                </form>
                            </li>
                        </ul>
                    </div>
                </li>
            );
        });

        return assemblyItemListJsx;
    }

    _buildAccordionListClassnames() {
        return classnames({
            vlist: true,
            'vlist-short': true,
            'u-isVisuallyHidden': !this.props.isOpen,
        });
    }

    render() {
        return (
            <li key={`assembly-${this.props.assembly.name}`}>
                {this._buildAssemblyGroupHeaderJsx()}
                <ul className={this._buildAccordionListClassnames()}>
                    {this._buildAssemblyItemList()}
                </ul>
            </li>
        );
    }
}

AssemblyItem.propTypes = {
    assembly: PropTypes.object.isRequired,
    assemblyItemsGroupedByAssemblyId: PropTypes.object.isRequired,
    itemIdMap: PropTypes.object.isRequired,
    departmentIdMap: PropTypes.object.isRequired,
    isOpen: PropTypes.bool.isRequired,
    onClickAccordionTriggerHandler: PropTypes.func.isRequired,
    onClickAddAssemblyHandler: PropTypes.func.isRequired,
    onClickAddAssemblyItemHandler: PropTypes.func.isRequired,
};
