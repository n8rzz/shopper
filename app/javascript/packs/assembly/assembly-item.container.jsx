import React from 'react';
import PropTypes from 'prop-types';
import ApiService from '../service/api.service';
import { AssemblyItem } from './assembly-item.component';

export class AssemblyItemContainer extends React.Component {
    constructor(props) {
        super(props);

        this._onClickAccordionTriggerHandler = this._onClickAccordionTrigger.bind(this);
        this._onClickAddAssemblyHandler = this._onClickAddAssembly.bind(this);
        this._onClickAddAssemblyItemHandler = this._onClickAddAssemblyItem.bind(this);

        this.state = {
            isOpen: false,
        };
    }

    _onClickAccordionTrigger() {
        this.setState((prevState) => ({ isOpen: !prevState.isOpen }));
    }

    _onClickAddAssembly(event) {
        event.preventDefault();

        console.log('+++ _onClickAddAssembly', this.props);

        // /order_items/create/assembly
        const assemblyId = this.props.assembly.id;
        // FIXME: move these away from data attributes
        const addAssemblyToOrderUrl = '/order_items/create/assembly.json';
        const requestPayload = {
            assembly_id: assemblyId,
        };

        ApiService.post(addAssemblyToOrderUrl, requestPayload, this.props.csrf)
            .then((response) => {
                if (response.status >= 300) {
                    console.error(`
                        Received an unexpected status code from ${addAssemblyToOrderUrl}. See response: ${response}
                    `);

                    return;
                }

                console.log('Assembly added successfully.  Need to trigger Notice banner');
                // this._onAddAssemblySuccess(requestPayload, response);
            })
            .catch((error) => { throw error; });
    }

    _onClickAddAssemblyItem(event) {
        event.preventDefault();

        const assemblyId = this.props.assembly.id;
        // FIXME: move these away from data attributes
        const { assemblyItemId, departmentId } = event.currentTarget.dataset;
        const assemblyItemUrl = '/order_items/create/item.json';
        const requestPayload = {
            assembly_id: assemblyId,
            department_id: departmentId,
            item_id: assemblyItemId,
        };

        ApiService.post(assemblyItemUrl, requestPayload, this.props.csrf)
            .then((response) => {
                if (response.status >= 300) {
                    console.error(`
                        Received an unexpected status code from ${assemblyItemUrl}. See response: ${response}
                    `);

                    return;
                }

                console.log('Item added successfully.  Need to trigger Notice banner');
                // this._onAddItemSuccess(requestPayload, response);
            })
            .catch((error) => { throw error; });
    }

    render() {
        return (
            <AssemblyItem
                assembly={this.props.assembly}
                assemblyItemsGroupedByAssemblyId={this.props.assemblyItemsGroupedByAssemblyId}
                departmentIdMap={this.props.departmentIdMap}
                itemIdMap={this.props.itemIdMap}
                isOpen={this.state.isOpen}
                onClickAccordionTriggerHandler={this._onClickAccordionTriggerHandler}
                onClickAddAssemblyHandler={this._onClickAddAssemblyHandler}
                onClickAddAssemblyItemHandler={this._onClickAddAssemblyItemHandler}
            />
        );
    }
}

AssemblyItemContainer.propTypes = {
    assembly: PropTypes.object.isRequired,
    assemblyItemsGroupedByAssemblyId: PropTypes.object.isRequired,
    csrf: PropTypes.string.isRequired,
    departmentIdMap: PropTypes.object.isRequired,
    itemIdMap: PropTypes.object.isRequired,
};
