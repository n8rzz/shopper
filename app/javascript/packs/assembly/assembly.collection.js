import AssemblyView from "./assembly.view";

export default class AssemblyCollection {
    get length() {
        return this._items.length;
    }

    constructor(elementList) {
        this._items = [];

        return this._init(elementList);
    }

    _init(elementList) {
        for (let i = 0; i < elementList.length; i++) {
            const element = elementList[i];
            const viewModel = new AssemblyView(element);

            this._add(viewModel);
        }

        return this;
    }

    _add(viewModel) {
        this._items.push(viewModel);
    }
}
