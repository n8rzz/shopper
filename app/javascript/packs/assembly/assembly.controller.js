// import ApiService from '../service/api.service';
import AssemblyCollection from './assembly.collection';

const CLASSNAMES = {
    ROOT_ELEMENT: 'js-assemblyList',
    TRIGGER_ELEMENT: 'js-assemblyList-trigger',
};

export default class AssemblyController {
    constructor() {
        const $element = document.getElementsByClassName(CLASSNAMES.TRIGGER_ELEMENT);
        this._collection = new AssemblyCollection($element);
        this._isEnabled = false;

        if ($element.length === 0) {
            return this.disable();
        }

        return this._init()
            ._createChildren()
            ._setupHandlers()
            .enable();
    }

    _init() {
        return this;
    }

    _createChildren() {
        return this;
    }

    _setupHandlers() {
        return this;
    }

    enable() {
        if (this._isEnabled) {
            return this;
        }

        this._isEnabled = true;

        return this;
    }

    disable() {
        if (!this._isEnabled) {
            return this;
        }

        this._isEnabled = false;

        return this;
    }
}

document.addEventListener('turbolinks:load', () => {
    const assemblyController = new AssemblyController();
});
