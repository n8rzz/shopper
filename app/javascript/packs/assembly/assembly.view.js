const CLASSNAME = {
    ASSEMBLY_ITEM_LIST: 'js-asemblyItemList',
    IS_VISUALLY_HIDDEN: 'u-isVisuallyHidden'
};

export default class AssemblyView {
    constructor($element) {
        const elementId = $element.dataset.assemblyItemId;
        this._$element = $element;
        this._$assemblyItemList = document.getElementsByClassName(`${CLASSNAME.ASSEMBLY_ITEM_LIST}-${elementId}`)[0];

        return this._init()
            ._createChildren()
            ._setupHandlers()
            .enable();
    }

    _init() {
        return this;
    }

    _createChildren() {
        return this;
    }

    _setupHandlers() {
        this._onClickHandler = this._onClick.bind(this);

        return this;
    }

    enable() {
        if (this._isEnabled) {
            return;
        }

        this._$element.addEventListener('click', this._onClickHandler);

        this._isEnabled = true;

        return this;
    }

    disable() {
        if (!this._isEnabled) {
            return;
        }

        this._$element.removeEventListener('click', this._onClickHandler);

        this._isEnabled = false;

        return this;
    }

    _onClick(event) {
        if (!this._hasAssemblyItemList()) {
            return;
        }

        if (!this._isHidden()) {
            this._$assemblyItemList.classList.add(CLASSNAME.IS_VISUALLY_HIDDEN);

            return;
        }

        this._$assemblyItemList.classList.remove(CLASSNAME.IS_VISUALLY_HIDDEN);
    }

    _hasAssemblyItemList() {
        return this._$assemblyItemList != null;
    }

    _isHidden() {
        return this._$assemblyItemList.classList.contains(CLASSNAME.IS_VISUALLY_HIDDEN);
    }
}
