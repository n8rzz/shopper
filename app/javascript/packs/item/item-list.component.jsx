import React from 'react';
import PropTypes from 'prop-types';
import { Item } from './item.component';

export function ItemList(props) {
    if (props.itemList.length === 0) {
        return null;
    }

    return (
        <ul className={'vlist'}>
            {props.itemList.map((item, index) => (
                <Item
                    csrf={props.csrf}
                    departmentId={item.department_id}
                    departmentName={props.departmentMap[item.department_id].name}
                    itemId={item.id}
                    itemName={item.name}
                    key={`item-${index}`}
                />
            ))}
        </ul>
    );
}

ItemList.propTypes = {
    csrf: PropTypes.string.isRequired,
    departmentMap: PropTypes.object.isRequired,
    itemList: PropTypes.array.isRequired,
};
