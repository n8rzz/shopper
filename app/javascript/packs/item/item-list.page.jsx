import React from 'react';
import PropTypes from 'prop-types';
import _groupBy from 'lodash/groupBy';
import _sortBy from 'lodash/sortBy';
import {FilterBar} from '../components/filter-bar/filter-bar.component';
import {ItemList} from './item-list.component';
import { GroupedItemList } from './grouped-item-list.component';

// TODO: abstract these constants to another file
export const FILTER_CONCERN = {
    ALPHA: 'ALPHA',
    DEPARTMENT: 'DEPARTMENT',
    FILTER: 'FILTER',
};
export const FILTER_LABEL = {
    [FILTER_CONCERN.ALPHA]: 'Alpha',
    [FILTER_CONCERN.DEPARTMENT]: 'Department',
    [FILTER_CONCERN.FILTER]: 'Filter',
}
export const FILTER_BAR_ITEMS = [
    {
        text: FILTER_LABEL.ALPHA,
        url: '',
    },
    {
        text: FILTER_LABEL.DEPARTMENT,
        url: '',
    },
    {
        text: FILTER_LABEL.FILTER,
        url: '',
    },
];
export const DEFAULT_DRILL_DOWN_VALUE = '--';

export class ItemListPage extends React.Component {
    constructor(props) {
        super(props);

        this._departmentMap = this._buildDepartmentMap(props);
        this._itemsByFirstLetter = this._buildItemsByFirstLetter(props);
        this._filterableItemList = Object.keys(this._itemsByFirstLetter);
        this.state = {
            concern: FILTER_CONCERN.ALPHA,
            itemsByFilter: props.items,
        };
    }

    componentWillMount() {
        this._onChangeFilterHandler = this._onChangeFilter.bind(this);
        this._onChangeLetterFilterValueHandler = this._onChangeLetterFilterValue.bind(this);
    }

    componentWillUnmount() {
        this._onChangeFilterHandler = null;
        this._onChangeLetterFilterValueHandler = null;
    }

    _buildSecondaryFilterOptionListJsx() {
        return (
            <select
                className="input-select"
                onChange={this._onChangeLetterFilterValueHandler}
            >
                <option value={DEFAULT_DRILL_DOWN_VALUE}>{DEFAULT_DRILL_DOWN_VALUE}</option>
                {this._filterableItemList.map((filterableLetter, index) => {
                    return (
                        <option value={filterableLetter} key={`select-${index}`}>
                            {filterableLetter.toUpperCase()}
                        </option>
                    );
                })}
            </select>
        );
    }

    _buildItemListForActiveConcernJsx() {
        switch (this.state.concern) {
            case FILTER_CONCERN.DEPARTMENT:
                return (
                    <GroupedItemList
                        csrf={this.props.csrf}
                        departments={this.props.departments}
                        departmentMap={this._departmentMap}
                        items={this.props.items}
                    />
                );
            case FILTER_CONCERN.FILTER:
                return (
                    <React.Fragment>
                        <div className="vr_3">
                            {this._buildSecondaryFilterOptionListJsx()}
                        </div>

                        <ItemList
                            csrf={this.props.csrf}
                            departmentMap={this._departmentMap}
                            itemList={this.state.itemsByFilter}
                        />
                    </React.Fragment>
                );
            default:
                return (
                    <ItemList
                        csrf={this.props.csrf}
                        departmentMap={this._departmentMap}
                        itemList={this.props.items}
                    />
                );
        }
    }

    render() {
        return (
            <React.Fragment>
                <FilterBar
                    items={FILTER_BAR_ITEMS}
                    onChangeFilterHandler={this._onChangeFilterHandler}
                />

                {this._buildItemListForActiveConcernJsx()}
            </React.Fragment>
        );
    }

    // TODO: rename this
    _onChangeLetterFilterValue(event) {
        const nextLetterFilter = event.target.value;
        let itemsByFilter = this._itemsByFirstLetter[nextLetterFilter];

        if (nextLetterFilter === DEFAULT_DRILL_DOWN_VALUE) {
            itemsByFilter = this.props.items;
        }

        this.setState({ itemsByFilter });
    }

    _onChangeFilter(event) {
        event.preventDefault();

        const filterName = event.currentTarget.text.toUpperCase();

        this.setState({ concern: FILTER_CONCERN[filterName] });
    }

    _buildDepartmentMap(props) {
        return props.departments.reduce((sum, department) => {
            sum[department.id] = department;

            return sum;
        }, {});
    }

    _buildItemsByFirstLetter(props) {
        return props.items.reduce((sum, item) => {
            const firstLetter = item.name.substr(0, 1).toLowerCase();

            if (typeof sum[firstLetter] !== 'undefined') {
                sum[firstLetter] = [
                    ...sum[firstLetter],
                    item,
                ];

                return sum;
            }

            sum[firstLetter] = [item];

            return sum;
        }, {});
    }
}

ItemListPage.propTypes = {
    csrf: PropTypes.string.isRequired,
    departments: PropTypes.array,
    items: PropTypes.array,
};
