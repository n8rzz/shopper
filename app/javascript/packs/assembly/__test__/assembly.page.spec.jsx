import React from 'react';
import { shallow } from 'enzyme';

import { AssemblyPage } from '../assembly.page';
import { assemblyListMock } from '../../__mocks__/assembly.mock';
import { csrfMock } from '../../__mocks__/csrf.mock';
import { departmentListMock } from '../../__mocks__/department.mock';
import { itemListMock } from '../../__mocks__/item.mock';

describe('AssemblyPage', () => {
    test('renders correctly with valid props', () => {
        const component = shallow(<AssemblyPage
            assemblies={assemblyListMock}
            assemblyItems={[]}
            csrf={csrfMock}
            departments={departmentListMock}
            items={itemListMock}
        />);

        expect(component).toMatchSnapshot();
    });
});
