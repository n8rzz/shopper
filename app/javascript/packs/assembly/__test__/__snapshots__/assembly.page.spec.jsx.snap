// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AssemblyPage renders correctly with valid props 1`] = `
<AssemblyItemList
  assemblies={
    Array [
      Object {
        "created_at": "2019-07-29T22:49:47.431Z",
        "id": 1,
        "name": "Souvlaki",
        "updated_at": "2019-07-29T22:49:47.431Z",
      },
      Object {
        "created_at": "2019-07-29T22:49:47.519Z",
        "id": 2,
        "name": "Chilli con Carne",
        "updated_at": "2019-07-29T22:49:47.519Z",
      },
    ]
  }
  assemblyItemsGroupedByAssemblyId={Object {}}
  csrf="3yIq75tyV791TFkzSZo/S4b+TVa/N5J3dI14PzMUPviGViEGeDSnGnvNix+MaB0RB0XuV7bpX8I3SmIEAr4/zQ=="
  departmentIdMap={
    Object {
      "3": Object {
        "created_at": "2019-07-21T16:03:32.883Z",
        "id": 3,
        "name": "Bakery",
        "updated_at": "2019-07-21T16:03:32.883Z",
      },
      "5": Object {
        "created_at": "2019-07-21T16:03:32.883Z",
        "id": 5,
        "name": "Meat",
        "updated_at": "2019-07-21T16:03:32.883Z",
      },
    }
  }
  itemIdMap={
    Object {
      "13": Object {
        "created_at": "2019-07-21T16:03:33.162Z",
        "department_id": 3,
        "id": 13,
        "name": "Agave Syrup",
        "updated_at": "2019-07-21T16:03:33.162Z",
      },
      "15": Object {
        "created_at": "2019-07-21T16:03:33.162Z",
        "department_id": 5,
        "id": 15,
        "name": "Agave Syrup",
        "updated_at": "2019-07-21T16:03:33.162Z",
      },
    }
  }
/>
`;
